### PR TITLE
UCP: TM offload sync send support

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe pls review

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1798/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3740/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
bot:ornl:retest

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3746/ for details (Mellanox internal link).

File: src/ucp/tag/offload.c [hunk 1]
```diff
@@ -371,15 +375,86 @@ const ucp_proto_t ucp_tag_offload_proto = {
...
+    ucp_offload_ssend_hdr_t *rep_hdr = dest;
+    ucp_request_t *req               = arg;
+
+    rep_hdr->sender_tag  = req->send.tag_offload.sender_tag;
+    rep_hdr->sender_uuid = req->send.tag_offload.sender_uuid;
+
+    return sizeof(*rep_hdr);
+}
+
+ucs_status_t ucp_offload_progress_sync_ack(uct_pending_req_t *self)
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: rename to ucp_tag_offload_progress_sync_ack

File: src/ucp/tag/offload.c [hunk 2]
```diff
@@ -371,15 +375,86 @@ const ucp_proto_t ucp_tag_offload_proto = {
...
+                                         uint64_t sender_uuid,
+                                         ucp_tag_t sender_tag)
+{
+    ucp_request_t *req;
+
+    ucs_trace_req("offload_send_sync_ack sender_uuid %"PRIx64" sender_tag %"PRIx64"",
+                  sender_uuid, sender_tag);
+
+    req = ucp_worker_allocate_reply(worker, sender_uuid);
+    req->send.uct.func                 = ucp_offload_progress_sync_ack;
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: can we reuse ucp_proto_progress_am_bcopy_single here?

File: src/ucp/core/ucp_types.h [hunk 3]
```diff
@@ -66,6 +66,7 @@ enum {
                                           rndv (bcopy) */
     UCP_AM_ID_RNDV_DATA_LAST    =  13, /* The last rndv data fragment when using
                                           software rndv (bcopy) */
+    UCP_AM_ID_OFFLOAD_SYNC_ACK  = 14,  /* Eager sync ack for tag offload proto */
```
Comment (yosefe) [inline src/ucp/core/ucp_types.h]: indent on column

File: src/ucp/tag/eager_rcv.c [hunk 4]
```diff
@@ -180,18 +212,34 @@ static ucs_status_t ucp_eager_sync_ack_handler(void *arg, void *data,
...
+                     UCP_RECV_DESC_FLAG_LAST  |
+                     UCP_RECV_DESC_FLAG_OFFLOAD;
+
+    if (ucs_unlikely(imm)) {
+        /* It is a sync send, imm data contains sender uuid */
+        hdr_len  = sizeof(ucp_eager_sync_hdr_t);
+        hdr      = (ucp_eager_hdr_t*)((char*)data - hdr_len);
+        sync_hdr = ucs_derived_of(hdr, ucp_eager_sync_hdr_t);
+        flags   |= UCP_RECV_DESC_FLAG_SYNC;
+        sync_hdr->req.reqptr      = 0ul;
```
Comment (yosefe) [inline src/ucp/tag/eager_rcv.c]: let's add assertion to sw (non-offload) sync ack flow, that reqptr!=0

File: src/ucp/tag/eager_rcv.c [hunk 5]
```diff
@@ -266,3 +314,5 @@ UCP_DEFINE_AM(UCP_FEATURE_TAG, UCP_AM_ID_EAGER_SYNC_FIRST, ucp_eager_sync_first_
               ucp_eager_dump, UCT_AM_CB_FLAG_SYNC);
 UCP_DEFINE_AM(UCP_FEATURE_TAG, UCP_AM_ID_EAGER_SYNC_ACK, ucp_eager_sync_ack_handler,
               ucp_eager_dump, UCT_AM_CB_FLAG_SYNC);
+UCP_DEFINE_AM(UCP_FEATURE_TAG, UCP_AM_ID_OFFLOAD_SYNC_ACK,
+              ucp_eager_offload_sync_ack_handler, ucp_eager_dump, UCT_AM_CB_FLAG_SYNC);
```
Comment (yosefe) [inline src/ucp/tag/eager_rcv.c]: need to add a case to the switch in ucp_eager_dump()

File: src/ucp/tag/offload.c [hunk 6]
```diff
@@ -371,15 +375,86 @@ const ucp_proto_t ucp_tag_offload_proto = {
...
+static ucs_status_t ucp_tag_offload_eager_sync_bcopy(uct_pending_req_t *self)
+{
+    ucp_request_t *req   = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_worker_t *worker = req->send.ep->worker;
+    ucs_status_t status;
+
+    status = ucp_do_tag_offload_bcopy(self, worker->uuid,
+                                      ucp_tag_offload_pack_eager);
+    if (status == UCS_OK) {
+        ucs_queue_push(&worker->context->tm.sync_reqs,
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: bug - need to push to queue before sending first fragment, since ack is sent for first fragment

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1843/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3784/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3818/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
jenkins passed but for some reason it does not update PR status

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3820/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
👍

**Pull request merged**